Mathlib/GroupTheory/DivisibleHull.lean

```diff
@@ -0,0 +1,462 @@
+
+theorem mk_eq_mk {m m' : M} {s s' : ℕ+} :
+    mk m s = mk m' s' ↔ ∃ u : ℕ+, u.val • s'.val • m = u.val • s.val • m' := by
+  unfold mk
+  rw [LocalizedModule.mk_eq, ↑ⁿ.exists_congr_left]
+  rfl
+
+/-- If `f : M → ℕ+ → α` respects the equivalence on localization,
+lift it to a function `DivisibleHull M → α`. -/
+def liftOn {α : Type*} (x : DivisibleHull M)
+    (f : M → ℕ+ → α)
+    (h : ∀ (m m' : M) (s s' : ℕ+), mk m s = mk m' s' → f m s = f m' s') : α :=
+  LocalizedModule.liftOn x (fun p ↦ f p.1 (↑ⁿ.symm p.2)) fun p p' heq ↦
+    h p.1 p'.1 (↑ⁿ.symm p.2) (↑ⁿ.symm p'.2) <| by
+      obtain ⟨u, hu⟩ := heq
+      exact mk_eq_mk.mpr ⟨↑ⁿ.symm u, hu⟩
+
+@[simp]
+theorem liftOn_mk {α : Type*} (m : M) (s : ℕ+)
+    (f : M → ℕ+ → α)
+    (h : ∀ (m m' : M) (s s' : ℕ+), mk m s = mk m' s' → f m s = f m' s') :
+    liftOn (mk m s) f h = f m s := rfl
+
+/-- If `f : M → ℕ+ → M → ℕ+ → α` respects the equivalence on
+localization, lift it to a function `DivisibleHull M → DivisibleHull M → α`. -/
+def liftOn₂ {α : Type*} (x y : DivisibleHull M)
+    (f : M → ℕ+ → M → ℕ+ → α)
+    (h : ∀ (m n m' n' : M) (s t s' t' : ℕ+),
+      mk m s = mk m' s' → mk n t = mk n' t' → f m s n t = f m' s' n' t') : α :=
+  LocalizedModule.liftOn₂ x y (fun p q ↦ f p.1 (↑ⁿ.symm p.2) q.1 (↑ⁿ.symm q.2))
+    fun p q p' q' heq heq' ↦
+    h p.1 q.1 p'.1 q'.1 (↑ⁿ.symm p.2) (↑ⁿ.symm q.2) (↑ⁿ.symm p'.2) (↑ⁿ.symm q'.2)
+      (by
+        obtain ⟨u, hu⟩ := heq
+        exact mk_eq_mk.mpr ⟨↑ⁿ.symm u, hu⟩)
+      (by
+        obtain ⟨u, hu⟩ := heq'
+        exact mk_eq_mk.mpr ⟨↑ⁿ.symm u, hu⟩)
+
+@[simp]
+theorem liftOn₂_mk {α : Type*} (m m' : M) (s s' : ℕ+)
+    (f : M → ℕ+ → M → ℕ+ → α)
+    (h : ∀ (m n m' n' : M) (s t s' t' : ℕ+),
+      mk m s = mk m' s' → mk n t = mk n' t' → f m s n t = f m' s' n' t') :
+    liftOn₂ (mk m s) (mk m' s') f h = f m s m' s' := rfl
+
+theorem mk_add_mk {m1 m2 : M} {s1 s2 : ℕ+} :
+    mk m1 s1 + mk m2 s2 = mk (s2.val • m1 + s1.val • m2) (s1 * s2) := LocalizedModule.mk_add_mk
+
+theorem mk_add_mk_left {m1 m2 : M} {s : ℕ+} :
+    mk m1 s + mk m2 s = mk (m1 + m2) s := by
+  rw [mk_add_mk, mk_eq_mk]
+  exact ⟨1, by simp [smul_smul]⟩
+
+@[simp]
+theorem coe_add {m1 m2 : M} : ↑(m1 + m2) = (↑m1 + ↑m2 : DivisibleHull M) := by simp [mk_add_mk_left]
+
+variable (M) in
+/-- Coercion from `M` to `DivisibleHull M` as a `AddMonoidHom`. -/
+@[simps]
+def coeAddMonoidHom : M →+ DivisibleHull M where
+  toFun := (↑)
+  map_zero' := by simp
+  map_add' := by simp
+
+theorem nsmul_mk (a : ℕ) (m : M) (s : ℕ+) : a • mk m s = mk (a • m) s := by
+  induction a with
+  | zero => simp
+  | succ n h => simp [add_nsmul, mk_add_mk_left, h]
+
+theorem nnqsmul_mk (a : ℚ≥0) (m : M) (s : ℕ+) :
+    a • mk m s = mk (a.num • m) (⟨a.den, a.den_pos⟩ * s) := by
+  convert LocalizedModule.mk'_smul_mk ℚ≥0 a.num m ⟨a.den, by simp⟩ (↑ⁿ s)
+  simp [IsLocalization.eq_mk'_iff_mul_eq]
+
+noncomputable
+instance : DivisibleBy (DivisibleHull M) ℕ where
```

Comment:
This instance is the main content of this PR, right? The rest is duplicating API to work with `DivisibleHull`, no?

Is there any chance you can shrink the `DivisibleHull` API to the very bare minimum, ie only the stuff that doesn't generalise to `LocalizedModule`? This instance in particular seems to generalise to
```suggestion
variable {M₀ : Type*} [CancelMonoidWithZero M₀]

instance : DivisibleBy (LocalizedModule (nonZeroDivisors M₀) M) M₀ where
```
Btw I find the order of the arguments to `DivisibleBy` quite surprising! Feel free to fix that in another PR

Reply:
I have removed the `DivisibleBy` part and added a note about it in the header doc.